src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -221,11 +266,11 @@ Object* FrozenObjectSegment::GetNextObject(Object* obj) const
 {
     // Input must not be null and should be within the segment
     _ASSERT(obj != nullptr);
-    _ASSERT((uint8_t*)obj >= m_pStart + sizeof(ObjHeader) && (uint8_t*)obj < m_pCurrent);
+    _ASSERT((uint8_t*)obj >= m_pStart + sizeof(ObjHeader) && (uint8_t*)obj < m_pCurrentRegistered);
 
     // FOH doesn't support objects with non-DATA_ALIGNMENT alignment yet.
```

Comment:
Do we need to set `m_NumComponents` for arrays as part of `TryAllocateObject`?

We are setting it too late and we can end up enumerating arrays without `m_NumComponents` that is not going to end wel..

Reply:
Good point! It also allowed to simplify the PublishObject logic a bit. The final API might be simplified a bit with C++ template to allow use of capturing lambdas for simplicity but that needed a bit more changes